# Bike Sharing Analytics


## Task


- **Question 1:** What are the most and least frequently rented hours?
- **Question 2:** In what season is bike rental peak?
- **Question 3:** How has the company's sales performance been in recent years?
- **Question 4:** How do registered customers compare to casual customers?
- **Question 5:** Prediction of bike rental count hourly or daily based on the environmental and seasonal settings with regression.
- **Question 6:** Event and Anomaly Detection, Count of rented bikes are also correlated to some events in the town which easily are traceable via search engines. For instance, query like "2012-10-30 washington d.c." in Google returns related results to Hurricane Sandy. Some of the important events are identified in [1]. Therefore the data can be used for validation of anomaly or event detection algorithms as well. [1] Fanaee-T, Hadi, and Gama, Joao, "Event labeling combining ensemble detectors and background knowledge", Progress in Artificial Intelligence (2013): pp. 1-15, Springer Berlin Heidelberg, doi:10.1007/s13748-013-0040-3.


## Library


In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import warnings
from datetime import time

warnings.filterwarnings("ignore")

## Data Wrangling


### Gathering Data


#### Daily data


In [2]:
# Load daily data
daily_df = pd.read_csv("./data/day.csv")
daily_df.head()

,instant,dteday,season,yr,mnth,holiday,weekday,workingday,weathersit,temp,atemp,hum,windspeed,casual,registered,cnt
0,1,2011-01-01,1,0,1,0,6,0,2,0.344167,0.363625,0.805833,0.160446,331,654,985
1,2,2011-01-02,1,0,1,0,0,0,2,0.363478,0.353739,0.696087,0.248539,131,670,801
2,3,2011-01-03,1,0,1,0,1,1,1,0.196364,0.189405,0.437273,0.248309,120,1229,1349
3,4,2011-01-04,1,0,1,0,2,1,1,0.200000,0.212122,0.590435,0.160296,108,1454,1562
4,5,2011-01-05,1,0,1,0,3,1,1,0.226957,0.229270,0.436957,0.186900,82,1518,1600


The dataset contains various columns, including:

- **instant**: Record index.
- **dteday**: Date of the record.
- **season**: The season in which the data was recorded (1: Spring, 2: Summer, 3: Fall, 4: Winter).
- **yr**: Year of the data (0: 2011, 1: 2012).
- **mnth**: Month of the year (1 to 12).
- **holiday**: Whether the day is a holiday (1: Yes, 0: No).
- **weekday**: Day of the week (0: Sunday, 6: Saturday).
- **workingday**: Whether the day is a working day (1: Yes, 0: No).
- **weathersit**: Weather situation (1: Clear, 2: Mist, 3: Light rain/snow, 4: Heavy rain/snow).
- **temp** : Normalized temperature in Celsius. The values are divided to 41 (max).
- **atemp**: Normalized feeling temperature in Celsius. The values are divided to 50 (max).
- **hum**: Normalized humidity. The values are divided to 100 (max).
- **windspeed**: Normalized wind speed. The values are divided to 67 (max).
- **casual**: Count of casual users.
- **registered**: Count of registered users.
- **cnt**: Count of total rental bikes including both casual and registered.

By examining the first few rows, we can get an overview of the dataset before performing further analysis.


#### Hourly data


In [3]:
# Load hour data
hourly_df = pd.read_csv("./data/hour.csv")
hourly_df.head()

,instant,dteday,season,yr,mnth,hr,holiday,weekday,workingday,weathersit,temp,atemp,hum,windspeed,casual,registered,cnt
0,1,2011-01-01,1,0,1,0,0,6,0,1,0.24,0.2879,0.81,0.0,3,13,16
1,2,2011-01-01,1,0,1,1,0,6,0,1,0.22,0.2727,0.80,0.0,8,32,40
2,3,2011-01-01,1,0,1,2,0,6,0,1,0.22,0.2727,0.80,0.0,5,27,32
3,4,2011-01-01,1,0,1,3,0,6,0,1,0.24,0.2879,0.75,0.0,3,10,13
4,5,2011-01-01,1,0,1,4,0,6,0,1,0.24,0.2879,0.75,0.0,0,1,1


This dataset contains similar features to the daily dataset (`day_df`), but with hourly records instead of daily aggregates. Some key columns include:

- **instant**: Unique identifier for each row.
- **dteday**: Date of the record.
- **hr**: Hour of the day (0 to 23).
- **season, yr, mnth, holiday, weekday, workingday, weathersit**: Similar to the daily dataset, representing seasonal and time-related information.
- **temp, atemp, hum, windspeed**: Weather-related features.
- **casual, registered, cnt**: Number of casual users, registered users, and total bike rentals for each hour.

Since this dataset provides a more granular view with hourly records, it allows for a deeper analysis of bike-sharing trends throughout the day.


### Assessing Data


#### Daily data


In [4]:
# Day data information
daily_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 731 entries, 0 to 730
Data columns (total 16 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   instant     731 non-null    int64  
 1   dteday      731 non-null    object 
 2   season      731 non-null    int64  
 3   yr          731 non-null    int64  
 4   mnth        731 non-null    int64  
 5   holiday     731 non-null    int64  
 6   weekday     731 non-null    int64  
 7   workingday  731 non-null    int64  
 8   weathersit  731 non-null    int64  
 9   temp        731 non-null    float64
 10  atemp       731 non-null    float64
 11  hum         731 non-null    float64
 12  windspeed   731 non-null    float64
 13  casual      731 non-null    int64  
 14  registered  731 non-null    int64  
 15  cnt         731 non-null    int64  
dtypes: float64(4), int64(11), object(1)
memory usage: 91.5+ KB


- The dataset contains **731 rows** and **16 columns**.
- All columns have **731 non-null values**, meaning there are **no missing values**.
- The dataset consists of different data types:
  - **Integer (`int64`)**: Used for categorical and count-based features such as `season`, `yr`, `mnth`, `holiday`, `weekday`, `workingday`, `weathersit`, `casual`, `registered`, and `cnt`.
  - **Float (`float64`)**: Used for continuous numerical values like `temp`, `atemp`, `hum`, and `windspeed`.
  - **Object (`object`)**: The `dteday` column, which represents dates in string format.

This information helps in preprocessing and deciding if any columns need transformation (e.g., converting `dteday` to a proper datetime format).


In [5]:
# Check Duplicate the day data
daily_df.duplicated().sum()

np.int64(0)

The output is **0**, which means there are no duplicate records in the dataset. This confirms that each row represents a unique entry, and no additional cleaning is required for duplicates.


In [6]:
# Check Missing Value the day data
daily_df.isna().sum()

instant       0
dteday        0
season        0
yr            0
mnth          0
holiday       0
weekday       0
workingday    0
weathersit    0
temp          0
atemp         0
hum           0
windspeed     0
casual        0
registered    0
cnt           0
dtype: int64

The output shows that **all columns have 0 missing values**, meaning the dataset is complete and does not require imputation or removal of missing data.

Since there are no missing values, we can proceed with the analysis without additional preprocessing for missing data.


In [7]:
# The descriptive statistic from day data
daily_df.describe()

,instant,season,yr,mnth,holiday,weekday,workingday,weathersit,temp,atemp,hum,windspeed,casual,registered,cnt
count,731.000000,731.000000,731.000000,731.000000,731.000000,731.000000,731.000000,731.000000,731.000000,731.000000,731.000000,731.000000,731.000000,731.000000,731.000000
mean,366.000000,2.496580,0.500684,6.519836,0.028728,2.997264,0.683995,1.395349,0.495385,0.474354,0.627894,0.190486,848.176471,3656.172367,4504.348837
std,211.165812,1.110807,0.500342,3.451913,0.167155,2.004787,0.465233,0.544894,0.183051,0.162961,0.142429,0.077498,686.622488,1560.256377,1937.211452
min,1.000000,1.000000,0.000000,1.000000,0.000000,0.000000,0.000000,1.000000,0.059130,0.079070,0.000000,0.022392,2.000000,20.000000,22.000000
25%,183.500000,2.000000,0.000000,4.000000,0.000000,1.000000,0.000000,1.000000,0.337083,0.337842,0.520000,0.134950,315.500000,2497.000000,3152.000000
50%,366.000000,3.000000,1.000000,7.000000,0.000000,3.000000,1.000000,1.000000,0.498333,0.486733,0.626667,0.180975,713.000000,3662.000000,4548.000000
75%,548.500000,3.000000,1.000000,10.000000,0.000000,5.000000,1.000000,2.000000,0.655417,0.608602,0.730209,0.233214,1096.000000,4776.500000,5956.000000
max,731.000000,4.000000,1.000000,12.000000,1.000000,6.000000,1.000000,3.000000,0.861667,0.840896,0.972500,0.507463,3410.000000,6946.000000,8714.000000


To understand the overall distribution of numerical variables, we use the `.describe()` function. This provides key statistical summaries, including:

- **count**: Number of non-null values in each column.
- **mean**: Average value of each column.
- **std (standard deviation)**: Measures data dispersion from the mean.
- **min and max**: Minimum and maximum values.
- **25%, 50% (median), and 75%**: Quartiles, showing data distribution.

Key Observations:

- The dataset consists of **731 records**.
- The **average number of total bike rentals (`cnt`) is 4504**, with a **minimum of 22** and a **maximum of 8714**.
- The **temperature (`temp`) ranges from 0.06 to 0.86** (normalized values).
- The dataset is **balanced across years (`yr`)**, with values of `0` (2011) and `1` (2012).
- The **working day (`workingday`) and holiday (`holiday`) columns are binary**, indicating whether a day is a working day or holiday.

This summary helps in understanding the overall trends and potential data distributions before further analysis.


#### Hourly data


In [8]:
# Hour data information
hourly_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 17379 entries, 0 to 17378
Data columns (total 17 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   instant     17379 non-null  int64  
 1   dteday      17379 non-null  object 
 2   season      17379 non-null  int64  
 3   yr          17379 non-null  int64  
 4   mnth        17379 non-null  int64  
 5   hr          17379 non-null  int64  
 6   holiday     17379 non-null  int64  
 7   weekday     17379 non-null  int64  
 8   workingday  17379 non-null  int64  
 9   weathersit  17379 non-null  int64  
 10  temp        17379 non-null  float64
 11  atemp       17379 non-null  float64
 12  hum         17379 non-null  float64
 13  windspeed   17379 non-null  float64
 14  casual      17379 non-null  int64  
 15  registered  17379 non-null  int64  
 16  cnt         17379 non-null  int64  
dtypes: float64(4), int64(12), object(1)
memory usage: 2.3+ MB


To understand the structure of the hourly dataset, we use the `.info()` function. This provides an overview of the dataset, including:

- **Total entries**: 17,379 rows.
- **Total columns**: 17 attributes.
- **Data types**:
  - `int64`: 12 columns (e.g., `season`, `yr`, `mnth`, `hr`, `holiday`, etc.).
  - `float64`: 4 columns (`temp`, `atemp`, `hum`, `windspeed`).
  - `object`: 1 column (`dteday`, representing dates).
- **Memory usage**: 2.3+ MB.

Key Observations:

- The dataset contains **hourly records of bike rentals** over two years.
- There are **no missing values** in any column, which simplifies data preprocessing.
- The `hr` column represents hours of the day (0-23), allowing for time-based analysis.
- `cnt` is the total number of rentals, obtained from the sum of `casual` and `registered` users.

This summary helps in understanding the dataset before performing further exploratory data analysis.


In [9]:
# Check duplicate the hour data
hourly_df.duplicated().sum()

np.int64(0)

The output is **0**, meaning there are **no duplicate entries** in the dataset. This confirms that each row represents a unique hourly record without redundancy.


In [10]:
# Check Missing Value the hour data
hourly_df.isnull().sum()

instant       0
dteday        0
season        0
yr            0
mnth          0
hr            0
holiday       0
weekday       0
workingday    0
weathersit    0
temp          0
atemp         0
hum           0
windspeed     0
casual        0
registered    0
cnt           0
dtype: int64

The output shows **0 missing values** in all 17 columns. This means the dataset is **fully complete**, with no need for imputation or data cleaning related to missing values.


In [11]:
# The descriptive statistic from day data
hourly_df.describe().round(2)

,instant,season,yr,mnth,hr,holiday,weekday,workingday,weathersit,temp,atemp,hum,windspeed,casual,registered,cnt
count,17379.00,17379.00,17379.0,17379.00,17379.00,17379.00,17379.00,17379.00,17379.00,17379.00,17379.00,17379.00,17379.00,17379.00,17379.00,17379.00
mean,8690.00,2.50,0.5,6.54,11.55,0.03,3.00,0.68,1.43,0.50,0.48,0.63,0.19,35.68,153.79,189.46
std,5017.03,1.11,0.5,3.44,6.91,0.17,2.01,0.47,0.64,0.19,0.17,0.19,0.12,49.31,151.36,181.39
min,1.00,1.00,0.0,1.00,0.00,0.00,0.00,0.00,1.00,0.02,0.00,0.00,0.00,0.00,0.00,1.00
25%,4345.50,2.00,0.0,4.00,6.00,0.00,1.00,0.00,1.00,0.34,0.33,0.48,0.10,4.00,34.00,40.00
50%,8690.00,3.00,1.0,7.00,12.00,0.00,3.00,1.00,1.00,0.50,0.48,0.63,0.19,17.00,115.00,142.00
75%,13034.50,3.00,1.0,10.00,18.00,0.00,5.00,1.00,2.00,0.66,0.62,0.78,0.25,48.00,220.00,281.00
max,17379.00,4.00,1.0,12.00,23.00,1.00,6.00,1.00,4.00,1.00,1.00,1.00,0.85,367.00,886.00,977.00


To understand the distribution of numerical features, we use the `.describe().round(2)` function. This provides key statistical summaries, including count, mean, standard deviation, min, max, and quartiles.

Key Observations:

- **Total Entries**: 17,379 rows.
- **Hourly Distribution** (`hr`): The values range from **0 to 23**, covering all hours in a day.
- **Temperature (`temp`) & Feels-Like Temperature (`atemp`)**:
  - `temp` (Normalized between 0 and 1) has a **mean of 0.50**, suggesting an even spread over different weather conditions.
  - `atemp` follows a similar distribution with a mean of **0.48**.
- **Humidity (`hum`) & Wind Speed (`windspeed`)**:
  - Humidity ranges from **0 to 1**, with a mean of **0.63**.
  - Wind speed has an average of **0.19**, indicating most days experience mild winds.
- **Bike Rentals (`cnt`)**:
  - The total count (`cnt`) ranges from **1 to 977** per hour.
  - Median rentals per hour is **142**, while the upper quartile reaches **281**, indicating some peak hours with high demand.
  - `casual` riders (mean **35.68**) are significantly fewer than `registered` riders (mean **153.79**), suggesting that most users are registered members.

Conclusion:  
This statistical summary provides insight into the distribution and variability of bike rentals across different hours, weather conditions, and user types. These insights will guide further analysis and model building.


### Cleaning Data


#### Hourly data


In [12]:
hourly_df.head()

,instant,dteday,season,yr,mnth,hr,holiday,weekday,workingday,weathersit,temp,atemp,hum,windspeed,casual,registered,cnt
0,1,2011-01-01,1,0,1,0,0,6,0,1,0.24,0.2879,0.81,0.0,3,13,16
1,2,2011-01-01,1,0,1,1,0,6,0,1,0.22,0.2727,0.80,0.0,8,32,40
2,3,2011-01-01,1,0,1,2,0,6,0,1,0.22,0.2727,0.80,0.0,5,27,32
3,4,2011-01-01,1,0,1,3,0,6,0,1,0.24,0.2879,0.75,0.0,3,10,13
4,5,2011-01-01,1,0,1,4,0,6,0,1,0.24,0.2879,0.75,0.0,0,1,1


In [13]:
# Change the column names
hourly_column = {
    "instant": "index",
    "dteday": "date",
    "yr": "year",
    "mnth": "month",
    "hr": "hour",
    "temp": "temperature",
    "atemp": "feels_like_temperature",
    "hum": "humidity",
    "cnt": "count",
}

hourly_df.rename(columns=hourly_column, inplace=True)

In [14]:
# Fix the datetime data
hourly_df["date"] = pd.to_datetime(hourly_df["date"])

In [15]:
# Categorical season column
season = {1: "springer", 2: "summer", 3: "fall", 4: "winter"}

hourly_df["season"] = hourly_df["season"].replace(season)

In [16]:
# Year column
year = {0: 2011, 1: 2012}

hourly_df["year"] = hourly_df["year"].replace(year)

In [17]:
# Month column
month = {
    1: "January",
    2: "February",
    3: "March",
    4: "April",
    5: "May",
    6: "June",
    7: "July",
    8: "August",
    9: "September",
    10: "October",
    11: "November",
    12: "December",
}

hourly_df["month"] = hourly_df["month"].replace(month)

In [18]:
# Hour Column
hourly_df["hour"] = hourly_df["hour"].apply(lambda x: time(x, 0))

In [19]:
# Change holiday column to categorical values
holiday = {0: "No", 1: "Yes"}

hourly_df["holiday"] = hourly_df["holiday"].replace(holiday)

In [20]:
# Change coloumns weekdays from numerical to categorical
weekdays = {
    0: "Sunday",
    1: "Monday",
    2: "Tuesday",
    3: "Wednesday",
    4: "Thursday",
    5: "Friday",
    6: "Saturday",
}

hourly_df["weekday"] = hourly_df["weekday"].replace(weekdays)

In [21]:
# Change coloumns workingday from numerical to categorical
workingday = {0: "No", 1: "Yes"}

hourly_df["workingday"] = hourly_df["workingday"].replace(workingday)

In [22]:
# Change the weathersit from numerical to categorical
weathersit_labels = {
    1: "Clear",
    2: "Cloudy or Misty",
    3: "Light Rain/Snow",
    4: "Heavy Rain/Snow",
}

hourly_df["weathersit"] = hourly_df["weathersit"].map(weathersit_labels)

In [23]:
# Denormalized temperature column with multiple by 41
hourly_df["temperature"] = hourly_df["temperature"] * 41

In [24]:
# Denormalized feels_like_temperature column with multiple by 50
hourly_df["feels_like_temperature"] = hourly_df["feels_like_temperature"] * 50

In [25]:
# Denormalized humidity column with multiple by 100
hourly_df["humidity"] = hourly_df["humidity"] * 100

In [26]:
# Denormalized windspeed column with multiple by 67
hourly_df["windspeed"] = hourly_df["windspeed"] * 67

In [27]:
hourly_df.head()

,index,date,season,year,month,hour,holiday,weekday,workingday,weathersit,temperature,feels_like_temperature,humidity,windspeed,casual,registered,count
0,1,2011-01-01,springer,2011,January,00:00:00,No,Saturday,No,Clear,9.84,14.395,81.0,0.0,3,13,16
1,2,2011-01-01,springer,2011,January,01:00:00,No,Saturday,No,Clear,9.02,13.635,80.0,0.0,8,32,40
2,3,2011-01-01,springer,2011,January,02:00:00,No,Saturday,No,Clear,9.02,13.635,80.0,0.0,5,27,32
3,4,2011-01-01,springer,2011,January,03:00:00,No,Saturday,No,Clear,9.84,14.395,75.0,0.0,3,10,13
4,5,2011-01-01,springer,2011,January,04:00:00,No,Saturday,No,Clear,9.84,14.395,75.0,0.0,0,1,1


#### Daily data


In [28]:
daily_df.head()

,instant,dteday,season,yr,mnth,holiday,weekday,workingday,weathersit,temp,atemp,hum,windspeed,casual,registered,cnt
0,1,2011-01-01,1,0,1,0,6,0,2,0.344167,0.363625,0.805833,0.160446,331,654,985
1,2,2011-01-02,1,0,1,0,0,0,2,0.363478,0.353739,0.696087,0.248539,131,670,801
2,3,2011-01-03,1,0,1,0,1,1,1,0.196364,0.189405,0.437273,0.248309,120,1229,1349
3,4,2011-01-04,1,0,1,0,2,1,1,0.200000,0.212122,0.590435,0.160296,108,1454,1562
4,5,2011-01-05,1,0,1,0,3,1,1,0.226957,0.229270,0.436957,0.186900,82,1518,1600


In [29]:
# Change the column names
daily_columns = {
    "instant": "index",
    "dteday": "date",
    "yr": "year",
    "mnth": "month",
    "temp": "temperature",
    "atemp": "feels_like_temperature",
    "hum": "humidity",
    "cnt": "count",
}

daily_df.rename(columns=daily_columns, inplace=True)

In [30]:
# Fix the datetime data
daily_df["date"] = pd.to_datetime(daily_df["date"])

In [31]:
# Categorical season column
season = {1: "springer", 2: "summer", 3: "fall", 4: "winter"}

daily_df["season"] = daily_df["season"].replace(season)

In [32]:
# Year column
year = {0: 2011, 1: 2012}

daily_df["year"] = daily_df["year"].replace(year)

In [33]:
# Month column
month = {
    1: "January",
    2: "February",
    3: "March",
    4: "April",
    5: "May",
    6: "June",
    7: "July",
    8: "August",
    9: "September",
    10: "October",
    11: "November",
    12: "December",
}

daily_df["month"] = daily_df["month"].replace(month)

In [34]:
# Change holiday column to categorical values
holiday = {0: "No", 1: "Yes"}

daily_df["holiday"] = daily_df["holiday"].replace(holiday)

In [35]:
# Change coloumns weekdays from numerical to categorical
weekdays = {
    0: "Sunday",
    1: "Monday",
    2: "Tuesday",
    3: "Wednesday",
    4: "Thursday",
    5: "Friday",
    6: "Saturday",
}

daily_df["weekday"] = daily_df["weekday"].replace(weekdays)

In [36]:
# Change coloumns workingday from numerical to categorical
workingday = {0: "No", 1: "Yes"}

daily_df["workingday"] = daily_df["workingday"].replace(workingday)

In [37]:
# Change the weathersit from numerical to categorical
weathersit_labels = {
    1: "Clear",
    2: "Cloudy or Misty",
    3: "Light Rain/Snow",
    4: "Heavy Rain/Snow",
}

daily_df["weathersit"] = daily_df["weathersit"].map(weathersit_labels)

In [38]:
# Denormalized temperature column with multiple by 41
daily_df["temperature"] = daily_df["temperature"] * 41

In [39]:
# Denormalized feels_like_temperature column with multiple by 50
daily_df["feels_like_temperature"] = daily_df["feels_like_temperature"] * 50

In [40]:
# Denormalized humidity column with multiple by 100
daily_df["humidity"] = daily_df["humidity"] * 100

In [41]:
# Denormalized windspeed column with multiple by 67
daily_df["windspeed"] = daily_df["windspeed"] * 67

In [42]:
daily_df.head()

,index,date,season,year,month,holiday,weekday,workingday,weathersit,temperature,feels_like_temperature,humidity,windspeed,casual,registered,count
0,1,2011-01-01,springer,2011,January,No,Saturday,No,Cloudy or Misty,14.110847,18.18125,80.5833,10.749882,331,654,985
1,2,2011-01-02,springer,2011,January,No,Sunday,No,Cloudy or Misty,14.902598,17.68695,69.6087,16.652113,131,670,801
2,3,2011-01-03,springer,2011,January,No,Monday,Yes,Clear,8.050924,9.47025,43.7273,16.636703,120,1229,1349
3,4,2011-01-04,springer,2011,January,No,Tuesday,Yes,Clear,8.200000,10.60610,59.0435,10.739832,108,1454,1562
4,5,2011-01-05,springer,2011,January,No,Wednesday,Yes,Clear,9.305237,11.46350,43.6957,12.522300,82,1518,1600


## Exploratory Data Analysis (EDA)


#### Exploratory Daily data


In [43]:
daily_df

,index,date,season,year,month,holiday,weekday,workingday,weathersit,temperature,feels_like_temperature,humidity,windspeed,casual,registered,count
0,1,2011-01-01,springer,2011,January,No,Saturday,No,Cloudy or Misty,14.110847,18.18125,80.5833,10.749882,331,654,985
1,2,2011-01-02,springer,2011,January,No,Sunday,No,Cloudy or Misty,14.902598,17.68695,69.6087,16.652113,131,670,801
2,3,2011-01-03,springer,2011,January,No,Monday,Yes,Clear,8.050924,9.47025,43.7273,16.636703,120,1229,1349
3,4,2011-01-04,springer,2011,January,No,Tuesday,Yes,Clear,8.200000,10.60610,59.0435,10.739832,108,1454,1562
4,5,2011-01-05,springer,2011,January,No,Wednesday,Yes,Clear,9.305237,11.46350,43.6957,12.522300,82,1518,1600
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
726,727,2012-12-27,springer,2012,December,No,Thursday,Yes,Cloudy or Misty,10.420847,11.33210,65.2917,23.458911,247,1867,2114
727,728,2012-12-28,springer,2012,December,No,Friday,Yes,Cloudy or Misty,10.386653,12.75230,59.0000,10.416557,644,2451,3095
728,729,2012-12-29,springer,2012,December,No,Saturday,No,Cloudy or Misty,10.386653,12.12000,75.2917,8.333661,159,1182,1341
729,730,2012-12-30,springer,2012,December,No,Sunday,No,Clear,10.489153,11.58500,48.3333,23.500518,364,1432,1796


In [44]:
# numerical column values
num = daily_df.select_dtypes(
    exclude=["object", "category", "datetime"]
).columns.tolist()
daily_df[num]

,index,year,temperature,feels_like_temperature,humidity,windspeed,casual,registered,count
0,1,2011,14.110847,18.18125,80.5833,10.749882,331,654,985
1,2,2011,14.902598,17.68695,69.6087,16.652113,131,670,801
2,3,2011,8.050924,9.47025,43.7273,16.636703,120,1229,1349
3,4,2011,8.200000,10.60610,59.0435,10.739832,108,1454,1562
4,5,2011,9.305237,11.46350,43.6957,12.522300,82,1518,1600
...,...,...,...,...,...,...,...,...,...
726,727,2012,10.420847,11.33210,65.2917,23.458911,247,1867,2114
727,728,2012,10.386653,12.75230,59.0000,10.416557,644,2451,3095
728,729,2012,10.386653,12.12000,75.2917,8.333661,159,1182,1341
729,730,2012,10.489153,11.58500,48.3333,23.500518,364,1432,1796


In [45]:
# datetyme column values
num = daily_df.select_dtypes(include=["datetime"]).columns.tolist()
daily_df[num]

,date
0,2011-01-01
1,2011-01-02
2,2011-01-03
3,2011-01-04
4,2011-01-05
...,...
726,2012-12-27
727,2012-12-28
728,2012-12-29
729,2012-12-30


In [46]:
# Categorical Column values
cat = daily_df.select_dtypes(include=["object", "category"]).columns.tolist()
daily_df[cat]

,season,month,holiday,weekday,workingday,weathersit
0,springer,January,No,Saturday,No,Cloudy or Misty
1,springer,January,No,Sunday,No,Cloudy or Misty
2,springer,January,No,Monday,Yes,Clear
3,springer,January,No,Tuesday,Yes,Clear
4,springer,January,No,Wednesday,Yes,Clear
...,...,...,...,...,...,...
726,springer,December,No,Thursday,Yes,Cloudy or Misty
727,springer,December,No,Friday,Yes,Cloudy or Misty
728,springer,December,No,Saturday,No,Cloudy or Misty
729,springer,December,No,Sunday,No,Clear


In [47]:
# rented by season
daily_df.groupby("season")["count"].agg({"max", "min", "mean", "std"}).round(2)

,min,max,mean,std
season,,,,
fall,1115,8714,5644.30,1459.80
springer,431,7836,2604.13,1399.94
summer,795,8362,4992.33,1695.98
winter,22,8555,4728.16,1699.62


In [48]:
# rented by weekday
daily_df.groupby("weekday")["count"].agg({"sum", "mean", "count"}).round(2)

,mean,count,sum
weekday,,,
Friday,4690.29,104,487790
Monday,4338.12,105,455503
Saturday,4550.54,105,477807
Sunday,4228.83,105,444027
Thursday,4667.26,104,485395
Tuesday,4510.66,104,469109
Wednesday,4548.54,104,473048


In [49]:
# rented by month
daily_df.groupby("month")["count"].agg({"sum", "mean", "count"}).round(2)

,mean,count,sum
month,,,
April,4484.90,60,269094
August,5664.42,62,351194
December,3403.81,62,211036
February,2655.30,57,151352
January,2176.34,62,134933
July,5563.68,62,344948
June,5772.37,60,346342
March,3692.26,62,228920
May,5349.77,62,331686


In [50]:
# Rented by season
daily_df.groupby("season")["count"].agg({"sum", "mean", "count"}).round(2)

,mean,count,sum
season,,,
fall,5644.30,188,1061129
springer,2604.13,181,471348
summer,4992.33,184,918589
winter,4728.16,178,841613


In [51]:
# Sales performance by years
daily_df.groupby("year")["count"].agg({"sum", "mean", "count"}).round(2)

,mean,count,sum
year,,,
2011,3405.76,365,1243103
2012,5599.93,366,2049576


In [52]:
# registered customers compare to casual customers
daily_df.groupby("weekday").agg(
    {"casual": ["mean", "sum"], "registered": ["mean", "sum"]}
).round(2)

casual         registered        
              mean     sum       mean     sum
weekday                                      
Friday      752.29   78238    3938.00  409552
Monday      674.13   70784    3663.99  384719
Saturday   1465.26  153852    3085.29  323955
Sunday     1338.30  140521    2890.53  303506
Thursday    590.96   61460    4076.30  423935
Tuesday     556.18   57843    3954.48  411266
Wednesday   551.14   57319    3997.39  415729

#### Exploratory Hourly data


In [65]:
hourly_df

,index,date,season,year,month,hour,holiday,weekday,workingday,weathersit,temperature,feels_like_temperature,humidity,windspeed,casual,registered,count
0,1,2011-01-01,springer,2011,January,00:00:00,No,Saturday,No,Clear,9.84,14.395,81.0,0.0000,3,13,16
1,2,2011-01-01,springer,2011,January,01:00:00,No,Saturday,No,Clear,9.02,13.635,80.0,0.0000,8,32,40
2,3,2011-01-01,springer,2011,January,02:00:00,No,Saturday,No,Clear,9.02,13.635,80.0,0.0000,5,27,32
3,4,2011-01-01,springer,2011,January,03:00:00,No,Saturday,No,Clear,9.84,14.395,75.0,0.0000,3,10,13
4,5,2011-01-01,springer,2011,January,04:00:00,No,Saturday,No,Clear,9.84,14.395,75.0,0.0000,0,1,1
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
17374,17375,2012-12-31,springer,2012,December,19:00:00,No,Monday,Yes,Cloudy or Misty,10.66,12.880,60.0,11.0014,11,108,119
17375,17376,2012-12-31,springer,2012,December,20:00:00,No,Monday,Yes,Cloudy or Misty,10.66,12.880,60.0,11.0014,8,81,89
17376,17377,2012-12-31,springer,2012,December,21:00:00,No,Monday,Yes,Clear,10.66,12.880,60.0,11.0014,7,83,90
17377,17378,2012-12-31,springer,2012,December,22:00:00,No,Monday,Yes,Clear,10.66,13.635,56.0,8.9981,13,48,61


In [66]:
# numerical column values
num = hourly_df.select_dtypes(
    exclude=["object", "category", "datetime"]
).columns.tolist()
hourly_df[num]

,index,year,temperature,feels_like_temperature,humidity,windspeed,casual,registered,count
0,1,2011,9.84,14.395,81.0,0.0000,3,13,16
1,2,2011,9.02,13.635,80.0,0.0000,8,32,40
2,3,2011,9.02,13.635,80.0,0.0000,5,27,32
3,4,2011,9.84,14.395,75.0,0.0000,3,10,13
4,5,2011,9.84,14.395,75.0,0.0000,0,1,1
...,...,...,...,...,...,...,...,...,...
17374,17375,2012,10.66,12.880,60.0,11.0014,11,108,119
17375,17376,2012,10.66,12.880,60.0,11.0014,8,81,89
17376,17377,2012,10.66,12.880,60.0,11.0014,7,83,90
17377,17378,2012,10.66,13.635,56.0,8.9981,13,48,61


In [67]:
# datetyme column values
num = hourly_df.select_dtypes(include=["datetime"]).columns.tolist()
hourly_df[num]

,date
0,2011-01-01
1,2011-01-01
2,2011-01-01
3,2011-01-01
4,2011-01-01
...,...
17374,2012-12-31
17375,2012-12-31
17376,2012-12-31
17377,2012-12-31


## Visualization & Explanatory Analysis


#### **Question 1:** What are the most and least frequently rented hours?


In [55]:
most_frequently_rented_hours = pd.pivot_table(
    hourly_df, values="count", index="hour", aggfunc="mean"
)
most_frequently_rented_hours.rename(columns={"count": "mean"}, inplace=True)
most_frequently_rented_hours.round().sort_values(by="mean", ascending=False)

,mean
hour,
17:00:00,461.0
18:00:00,426.0
08:00:00,359.0
16:00:00,312.0
19:00:00,312.0
13:00:00,254.0
12:00:00,253.0
15:00:00,251.0
14:00:00,241.0


In [56]:
least_frequently_rented_hours = pd.pivot_table(
    hourly_df, values="count", index="hour", aggfunc="mean"
)
least_frequently_rented_hours.rename(columns={"count": "mean"}, inplace=True)
least_frequently_rented_hours.round().sort_values(by="mean", ascending=True)

,mean
hour,
04:00:00,6.0
03:00:00,12.0
05:00:00,20.0
02:00:00,23.0
01:00:00,33.0
00:00:00,54.0
06:00:00,76.0
23:00:00,88.0
22:00:00,131.0


In [57]:
most_frequently_rented_daily = pd.pivot_table(
    daily_df, values="count", index="weekday", aggfunc="mean"
)
most_frequently_rented_daily.rename(columns={"count": "mean"}, inplace=True)
most_frequently_rented_daily.round().sort_values(by="mean", ascending=False)

,mean
weekday,
Friday,4690.0
Thursday,4667.0
Saturday,4551.0
Wednesday,4549.0
Tuesday,4511.0
Monday,4338.0
Sunday,4229.0


In [58]:
least_frequently_rented_daily = pd.pivot_table(
    daily_df, values="count", index="weekday", aggfunc="mean"
)
least_frequently_rented_daily.rename(columns={"count": "mean"}, inplace=True)
least_frequently_rented_daily.round().sort_values(by="mean", ascending=True)

,mean
weekday,
Sunday,4229.0
Monday,4338.0
Tuesday,4511.0
Wednesday,4549.0
Saturday,4551.0
Thursday,4667.0
Friday,4690.0


#### **Question 2:** In what season is bike rental peak?


In [59]:
peak_season_rented = pd.pivot_table(
    daily_df, values="count", index="season", aggfunc="mean"
)
peak_season_rented.rename(columns={"count": "rented_mean"}, inplace=True)
peak_season_rented.round(2).sort_values(by="rented_mean", ascending=False)

,rented_mean
season,
fall,5644.30
summer,4992.33
winter,4728.16
springer,2604.13


In [60]:
peak_season_rented = pd.pivot_table(
    hourly_df, values="count", index="season", aggfunc="mean"
)
peak_season_rented.rename(columns={"count": "rented_mean"}, inplace=True)
peak_season_rented.round(2).sort_values(by="rented_mean", ascending=False)

,rented_mean
season,
fall,236.02
summer,208.34
winter,198.87
springer,111.11


#### **Question 3:** How has the company's sales performance been in recent years?


In [61]:
recent_performance_daily = daily_df.groupby("date")["count"].sum()

recent_performance_daily

date
2011-01-01     985
2011-01-02     801
2011-01-03    1349
2011-01-04    1562
2011-01-05    1600
              ... 
2012-12-27    2114
2012-12-28    3095
2012-12-29    1341
2012-12-30    1796
2012-12-31    2729
Name: count, Length: 731, dtype: int64

#### **Question 4:** How do registered customers compare to casual customers?


In [62]:
compare_casual_regitered = daily_df.groupby("weekday")[["registered", "casual"]].agg(
    ["mean", "sum", "count"]
)

compare_casual_regitered.round(2)

registered                 casual              
                mean     sum count     mean     sum count
weekday                                                  
Friday       3938.00  409552   104   752.29   78238   104
Monday       3663.99  384719   105   674.13   70784   105
Saturday     3085.29  323955   105  1465.26  153852   105
Sunday       2890.53  303506   105  1338.30  140521   105
Thursday     4076.30  423935   104   590.96   61460   104
Tuesday      3954.48  411266   104   556.18   57843   104
Wednesday    3997.39  415729   104   551.14   57319   104

In [63]:
compare_casual_regitered = hourly_df.groupby("hour")[["registered", "casual"]].agg(
    ["mean", "sum", "count"]
)

compare_casual_regitered.round(2)

registered               casual             
               mean     sum count   mean    sum count
hour                                                 
00:00:00      43.74   31755   726  10.16   7375   726
01:00:00      26.87   19455   724   6.50   4709   724
02:00:00      18.10   12940   715   4.77   3412   715
03:00:00       9.01    6281   697   2.72   1893   697
04:00:00       5.10    3554   697   1.25    874   697
05:00:00      18.48   13249   717   1.41   1012   717
06:00:00      71.88   52115   725   4.16   3017   725
07:00:00     201.01  146134   727  11.06   8037   727
08:00:00     337.33  245240   727  21.68  15761   727
09:00:00     188.42  136980   727  30.89  22458   727
10:00:00     127.19   92468   727  46.48  33789   727
11:00:00     148.60  108034   727  59.54  43286   727
12:00:00     185.02  134696   728  68.29  49718   728
13:00:00     181.35  132206   729  72.31  52713   729
14:00:00     165.38  120563   729  75.57  55089   729
15:00:00     176.33  128543   729  74.91  54606   729
16:00:00     238.24  173914   730  73.75  53834   730
17:00:00     387.18  282640   730  74.27  54220   730
18:00:00     364.39  265276   728  61.12  44496   728
19:00:00     262.75  191284   728  48.77  35505   728
20:00:00     189.80  138172   728  36.23  26378   728
21:00:00     144.06  104875   728  28.26  20570   728
22:00:00     109.08   79412   728  22.25  16200   728
23:00:00      72.63   52876   728  15.20  11065   728

#### **Question 5:** Prediction of bike rental count hourly or daily based on the environmental and seasonal settings with regression.


#### **Question 6:** Event and Anomaly Detection, Count of rented bikes are also correlated to some events in the town which easily are traceable via search engines. For instance, query like "2012-10-30 washington d.c." in Google returns related results to Hurricane Sandy. Some of the important events are identified in [1]. Therefore the data can be used for validation of anomaly or event detection algorithms as well. [1] Fanaee-T, Hadi, and Gama, Joao, "Event labeling combining ensemble detectors and background knowledge", Progress in Artificial Intelligence (2013): pp. 1-15, Springer Berlin Heidelberg, doi:10.1007/s13748-013-0040-3.


## Conclusion


- **Question 1:** What are the most and least frequently rented hours?
- **Question 2:** In what season is bike rental peak?
- **Question 3:** How has the company's sales performance been in recent years?
- **Question 4:** How do registered customers compare to casual customers?
- **Question 5:** Prediction of bike rental count hourly or daily based on the environmental and seasonal settings with regression.
- **Question 6:** Event and Anomaly Detection, Count of rented bikes are also correlated to some events in the town which easily are traceable via search engines. For instance, query like "2012-10-30 washington d.c." in Google returns related results to Hurricane Sandy. Some of the important events are identified in [1]. Therefore the data can be used for validation of anomaly or event detection algorithms as well. [1] Fanaee-T, Hadi, and Gama, Joao, "Event labeling combining ensemble detectors and background knowledge", Progress in Artificial Intelligence (2013): pp. 1-15, Springer Berlin Heidelberg, doi:10.1007/s13748-013-0040-3.
